### DISCUSSION TITLE
set_pipeline step modifies current pipeline, commits seppuku.

### DISCUSSION THREAD

**Discussion by xfin-dragontamer** [concourse/concourse#6784] [Help & Support]
I'm a member of a Comcast team which creates pipeline templates used by our division.  Yesterday, our concourse team updated us from concourse 6.4.1 to 6.5.1 and it broke our pipeline generation, so I'm here to dig for information on what the concourse developers' concept is on usage of the `set_pipeline` step.

The reason our pipeline gen broke is that we have been working around the fact that it's nearly impossible (in our environment) to automate authentication to concourse.  That means we can't have a service that can construct a pipeline.yml, auth to the concourse environment, and execute a set-pipeline.   

Instead, we have a pipeline which a team, wanting to generate a pipeline for their application, will issue a fly command with our pipeline template with a bunch of parameters targeting a specific application project.  The pipeline template reaches out and grabs information about the target app from several places, generates a pipeline.yml template custom made for the target service, then executes a `set_pipeline` task, on itself, with the new pipeline.yml.   The custom-made pipeline yml does not include another `set_pipeline` task.  

In the 6.5.1 update, this happened:
> Automatically archive pipelines set by a set_pipeline step that meets one of the following criteria: #5854
    - set_pipeline step is removed from job
    - Job that set pipeline is deleted
    - Parent pipeline is deleted

So now our pipeline template executes, custom tailors itself for the target service, regenerates itself with `set_pipeline`, then sets itself to archived.  Developers cry.  Managers yell.  Joy is lost.

Am I totally misreading what the purpose or potential usage of the `set_pipeline` task is supposed to be?  Or is my team the only one that thought of this scenario?

**Comment by aoldershaw** [comment]
There are two intended use-cases for the `set_pipeline` step instead of `fly set-pipeline`:

1. You want to continuously keep your pipeline up-to-date based on changes to some external source of truth (i.e. resource versions). e.g. you have a repo containing your pipelines, and a job which triggers on changes to the repo and runs `set_pipeline` for each one. Then, to update your pipelines, you'd make a `git commit` and Concourse would take care of the rest
2. (Eventually) manage a dynamically growing/shrinking group of pipelines. e.g. set a pipeline for each PR - when PRs are merged/closed, they should be archived. This use-case isn't ready, but we're working towards it.

I'm not sure I understand your use-case entirely, so I'll try to put it in my own words - tell me if my interpretation is off:

* An application team needs to create a pipeline
* You (manually?) set an initial "placeholder" pipeline in place of the actual pipeline they want
* The application team triggers that placeholder pipeline once with some configuration
  * This configuration is used to render the a pipeline template
  * The job then uses `set_pipeline` to replace the placeholder pipeline with the rendered one

Is that accurate?

If so, does this need to be a separate pipeline? I'm guessing you want to use Concourse to render the pipeline template - could you get away with a two-step process like:

```
fly execute ... -o template=./template
fly sp ... -c ./template/template.yml
```

Of course, that won't work so well if the automation is more than a `task` can handle on its own.

---

Follow-up question - what happens if the application team needs to update their pipeline (e.g. the configuration they initially provided to render the template has changed)? Assuming my initial interpretation was correct (there's a placeholder pipeline with a job, `render-template` say, which they run `fly trigger-job` on or something), what if the pipeline template kept that job intact? So, the pipeline may start out looking like:

```
┌───────────────────┐
│                   │
│  render-template  │
│                   │
└───────────────────┘
```

and after it sets itself, it looks like:

```
   ┌─────────────┐      ┌────────────┐
   │             │      │            │
   │    job1     ├──────►    job2    │
   │             │      │            │
   └─────────────┘      └────────────┘

┌───────────────────┐
│                   │
│  render-template  │
│                   │
└───────────────────┘
```

That way, the pipeline won't get auto-archived (since the `render-template` job still exists), and the team is able to update their pipeline manually using `fly trigger-job`.

**Reply by taylorsilva** [reply]
Adding to Aidan's excellent comment, the source of motivations is the archiving pipelines rfc: https://github.com/concourse/rfcs/blob/master/033-archiving-pipelines/proposal.md

**Reply by xfin-dragontamer** [reply]
I think our views of this problem domain are from two different viewpoints. I understand your use-cases.  Let me clarify my team's use cases.

Our applications are deployed to the cloud in banks of multiple instances.  Also, we can deploy multiple banks of an application to multiple data centers.  So for application "A", as an example, can have a 100 instance bank running in a data center on the east coast, and 100 instance bank running in a data center on the west coast.  If you take into account that we could deploy our applications to many data centers ("A" gets deployed to 6 different data centers), and every application may have a different mix of which data centers they deploy to, you start to see it's kind of complex here.  Now take into consideration that we have hundreds of applications.

So:
1. An application gets deployed to 1-to-many DCs
2. Every application has a different footprint of which DCs they deploy to
3. Each team owns multiple applications.

The problem we face is that we can't centralize our pipelines as you do in your use-case number 1.  The reason is our pipelines _are not static_, since there are too many different possible configurations for each team's pipelines.  Also, it's our goal to have our pipeline code be completely external to an application's source code.   The answer was one single morphic pipeline template that does all the heavy lifting of discovering an application's needs and custom-assembling a pipeline for each application.

To expound upon on our steps:
1. Teams check out our application pipeline template.
2. Team issues a fly command with parameters telling the pipeline the target application and where it's deployment values are, and then triggers the job.   
    _e.g._ pipeline "my-application:v1.1.0" is created with one "createPipeline" job.
2. The resulting pipeline runs a java library jar that uses the fly command parameter values to reach out to a number of internal systems and collect all the information about the target application.
3. The library jar then custom-assembles a pipeline for that application, creating jobs for each data center it deploys to, what kind of scans the code needs, etc...
4. The `set_pipeline` task runs, taking the generated pipeline.yml and a parameter yaml file, and runs on itself.  The effect of which is the single "createPipeline" job in the pipeline is replaced with a robust set of jobs that deploy to multiple environments and data centers.
   _e.g._ pipeline "my-application:v1.1.0" now has 10 jobs ("deployDev", "code-analysis", "deployQa", "createRelease", etc...)
5. Lastly webhooks for the pipeline's target github branch will kick off a pipeline run when the branch is updated.

Now, the idea of updating the pipeline dynamically is something that was on our radar.  Currently, a team will re-run the fly command to update the pipeline if they have to pick up new environments the application need to deploy to, or when we release a new version of the pipeline template.  We can't have a discreet "update" pipeline for every application pipeline we run because we already have over 17,000 pipelines running in our environment.  Even if only a third of those are applications, well, the math is pretty harsh for platform resource requirements.

Again, in our environment, we have no way to programmatically authenticate to Concourse, which in our case uses our Github Enterprise for auth.   If we could do that, we'd have a service that does the pipeline assembly and management.

**Reply by xfin-dragontamer** [reply]
@aoldershaw Does that clear anything up?

**Reply by aoldershaw** [reply]
Thanks for elaborating on your use case! Sounds like a very interesting setup. I think I mostly understand, but have a couple comments/questions.

The easiest way I can think of to "get around" the auto-archiving is for the `createPipeline` job to generate a pipeline yaml that contains the jobs it currently would create based on the configuration (e.g. `deployDev`, `code-analysis`, ...) AND the `createPipeline` job (like my diagram in the bottom of my first comment). That way, when it sets itself, the `createPipeline` job would still exist, so the pipeline wouldn't get auto-archived.

I put `"get around"` in quotes because I think it has some additional benefits - for instance, and this may not be possible depending on how the pipeline template works, but the `createPipeline` job could trigger based on changes to the pipeline template repo. Also...

> Team issues a fly command with parameters telling the pipeline the target application and where it's deployment values are, and then triggers the job.

When you say `where its deployment values are`, does this mean this configuration is stored in some external location? If so, you could potentially *also* trigger the `createPipeline` job when this configuration changes. So, when an application team wants to make a change to their configuration, they just update the configuration and the pipeline could (potentially) be updated automatically.

---

Apologies if I'm still not understanding - let me know if this suggestion isn't feasible.

**Reply by xfin-dragontamer** [reply]
@aoldershaw I really appreciate you taking the time to have this conversation.

I figured incorporating the `createPipeline` job back into the generated pipeline yml would fix this, but there are a few caveats.

1. Our pipeline template updates may have breaking changes in them.  The template may look for information in a new place that the team needs to set up first, or the template may support a newer version of Java, or such changes like that.  We can't force the teams to take our updates immediately.  Due to this, when we have a new version of our template, it's a new branch in our template git repository.
2. We have several versions of our pipeline templates in production at the same time.  Right now, we have teams that are still on Java 8, and so we have a branch of our pipeline templates for Java 8.  The teams that have moved to Java 11 are on newer branches of the pipeline templates.  The different versions of the templates may (probably) pull different docker images for the jobs.

This is the missing piece that explains how the last two points work:  Our templates refer to themselves in the created pipelines.  If you create a pipeline with the template github branch 1.0.0, for example, when the pipeline runs, that specific pipeline branch is a resource which is loaded into the job docker instance so the template shell scripts can be run.   I think this is a standard concourse pipeline setup?  This means that upgrading a pipeline is necessarily a manual operation, since the template knows it's own version number and to update a template, you have to check out the template branch you want to update to and run the fly command in that branch.

Regarding the deployment values - we use Cloud Foundry, so each application has deployment manifests for each environment which the application is deployed to. Those manifests (and the other value sources) are in standard places we dictate (configuration server, github repo, etc...). The "createPipeline" job jar library loads the manifests and other values from those known locations, and then creates deployment jobs for each manifest it finds and adjusts the rest of the pipeline yaml as necessary give the other values.

All this boils down to this:  We could build in the `createPipeline` job into the template (and maybe change the name to "updatePipeline", but it should only probably run when there's a change to the application's deployment manifests or such.  Is there a way to have an "optional" job?  I mean, a job that you can run only when needed, but not interfering with the rest of the pipeline?  That would probably be the ideal in this case.  The "updatePipeline" job only runs when the team is ready with their new manifests, and it wouldn't add more latency to a pipeline run time.

As an aside, I've also got this idea to use `set_pipeline` during a pipeline run, updating the job names with the version of the application currently being built.  I like bells and whistles.

Again most of our problems would probably go away if we could programmatically authenticate to concourse. If you have any resources about getting around authentication issues, I'd be mighty glad to see them.

**Reply by aoldershaw** [reply]
Fair points, it doesn't necessarily make sense to continuously bump the pipeline template - the point I'm trying to illustrate is that this model grants you some flexibility in updating the pipelines, since you *can* automate it if it's feasible to.

> Is there a way to have an "optional" job? I mean, a job that you can run only when needed, but not interfering with the rest of the pipeline?

Is that just a job without any `triggers` or `passed` constraints? When you want to update the pipeline, you can just manually trigger the `updatePipeline` job, and it'll set itself. You can also tuck it away in the UI by putting it in its own [`group`](https://concourse-ci.org/pipelines.html#schema.pipeline.groups), if that's what you want.

> Again most of our problems would probably go away if we could programmatically authenticate to concourse. If you have any resources about getting around authentication issues, I'd be mighty glad to see them.

I hope you don't feel I'm ignoring this point! People have often requested the ability to more easily authenticate with Concourse. However, we want to better understand the use cases for doing so - typically, it's to automate some aspect of Concourse itself. Given that Concourse *is* an automation engine, it tends to be an indication that there's some limitation in what Concourse is capable of, and that there's perhaps a feature yet to be created that'd solve the problem (and others like it) in a more general way.

Granted, that doesn't help with problems you have now. If you *really* need to authenticate with Concourse in automation, our typical suggestion is to setup a basic auth user - which obviously is far from a perfect solution, and may not be possible given your security requirements.

**Reply by xfin-dragontamer** [reply]
> the point I'm trying to illustrate is that this model grants you some flexibility in updating the pipelines, since you can automate it if it's feasible to.

Oh, I am definitely going to try adding the `updatePipeline` job to the constructed yaml. It would be nice if the teams only had to run a job to update their pipeline's deployment job list rather than having to re-run the fly command.   I will look into the "groups" thing too.   Thanks for that hint.

> However, we want to better understand the use cases for doing so - typically, it's to automate some aspect of Concourse itself. Given that Concourse is an automation engine, it tends to be an indication that there's some limitation in what Concourse is capable of, and that there's perhaps a feature yet to be created that'd solve the problem (and others like it) in a more general way.

I had to think about this one for a bit.

From just my point of view, our... issue? - it's not an issue, per se, but the challenge we face with our large end-user base is that concourse pipelines are created with a single set of fairly immutable values.  When you issue the fly command, you can pass in parameters, but those don't change for the life of the pipeline.  If you use something like Vault for some of the values, the pipeline retrieves them on every run, but systems like vault are a bit antagonistic to outside automated management themselves, admittedly in some cases by necessary design.   The inputs and outputs of the concourse jobs are also rather immutable.  It's not possible to create a pipeline that can run when any branch is created in a git repository - when the pipeline is created, your git resource is only going to point at one branch of one repository, and it probably wouldn't even be a good idea or feasible to do such a thing.

When it comes right down to it, there's going to be a need, in a large sprawling system like ours, for some kind of outside management tool to bridge the gap between someone creating a branch in github, and concourse spawning a pipeline for that branch creation.   By design, concourse can't handle mutability very well once a pipeline is created.  And that's okay, because what it does, it does really well.  We've been through Jenkins and GoCD and even with the huge startup cost of designing a pipeline, I'd take concourse over them any day because I can always design pipelines to serve our specific needs, and there's the crux of it.  We have need of many, many different types of pipelines and as far as I can see, you can't really create a concourse pipeline that can spawn multiple structures of pipelines, on demand, when a user wants, with different inputs every time.

Let me put it this way - it would be eminently easy for us to write a reactjs UI with a spring boot backend that can let a user select which git project, and which branch of that project, they want to create a pipeline for, add a few values, and hit a button.   That same management app could, with webhooks, automatically create and destroy pipelines when git repository branches with specific naming conventions are created and destroyed.  Given how concourse works (ASAIK), there's no way to do something like that _with_ a concourse pipeline.  You have automation down pat, and the resource plugins are adding more capability day by day.  It's just not in the DNA of concourse to be flexible enough to manage _itself_.  The main issue I keep coming back to in my mind is the lack of any way to consume user input.  Any ad-hoc input to a pipeline would have to be backdoored via a database resource or something, which, when you follow that trail, leads you right back to needing a UI that a user can input values to.

Anyway, I'll like to hear your thoughts about that.  If you'd ever like to do a zoom or such with our internal comcast concourse team, we would love to set that up.  You're probably already talking to us on some level, though.

**Reply by chenbh** [reply]
Would you be able to have the end-users make a commit/PR to some centralized repo which is then used to generate the final pipeline? You can have some master pipeline watching this repo and use the same templating logic to [set pipelines in other teams](https://concourse-ci.org/jobs.html#schema.step.set-pipeline-step.team).

This also has the benefit of having your config/deployment values stored as code. Concourse tries to encourage storing state outside of itself, one of the thoughts behind forcing pipelines to only be set through CLI instead of UI is that it encourages people to check in their pipeline.yml into their codebase. The idea of a metamorphosing pipeline kinda defeats this purpose since this pipeline config now only exists inside Concourse and becomes hard to version control.

We've noticed a pattern in very large enterprise deployments of Concourse where they have a dedicated "platform-operators" team for writing pipeline template for other developers. They usually provide some sort of self-service for end-users to bring their own config, which is then used to generate a fully featured pipeline.

I have some additional questions to about your problem space:

1. Are you part of the main team on your Concourse installation? If we can safely say most platform-operators are on the main team then we can look into adding more advanced features specific to that team.
2. Are you also in charge of team (end-user) creation/onboarding or is that another team's responsibility? 
3. Do the end-users have the ability to modify the final pipeline and is it common for them to do so?

While I don't think it'll fully solve your use case, you might be interested in the [projects RFC](https://github.com/concourse/rfcs/pull/32) which expands on the set_pipeline step
